sympy/physics/vector/point.py

```diff
@@ -483,19 +483,56 @@ def vel(self, frame):
         Examples
         ========
 
-        >>> from sympy.physics.vector import Point, ReferenceFrame
+        >>> from sympy.physics.vector import Point, ReferenceFrame, dynamicsymbols
         >>> N = ReferenceFrame('N')
         >>> p1 = Point('p1')
         >>> p1.set_vel(N, 10 * N.x)
         >>> p1.vel(N)
         10*N.x
+        >>> p = Point('p')
+        >>> q = dynamicsymbols('q')
+        >>> p.set_vel(N, 10 * N.x)
+        >>> p2 = Point('p2')
+        >>> p2.set_pos(p, q*N.x)
+        >>> p2.vel(N)
+        (Derivative(q(t), t) + 10)*N.x
 
         """
 
         _check_frame(frame)
         if not (frame in self._vel_dict):
-            raise ValueError('Velocity of point ' + self.name + ' has not been'
+            starting_node = self
+            visited = []
+            queue = [starting_node]
+            req_point = None
+            while queue: #BFS to find nearest point
+                node = queue.pop(0)
+                if node not in visited:
+                    visited.append(node)
+                    for neighbor, neighbor_pos in node._pos_dict.items():
+                        try:
+                            neighbor_pos.dt(frame) #Checks if pos vector is valid
+                        except ValueError:
+                            continue
+                        else:
+                            try :
+                                neighbor._vel_dict[frame] #Checks if point has its vel defined in req frame
+                            except KeyError:
+                                queue.append(neighbor)
+                            else:
+                                req_point = neighbor # Nearest valid point
+                                break
+                if req_point is not None:
+                        break
+            else:
+                raise ValueError('Velocity of point ' + self.name + ' has not been'
                              ' defined in ReferenceFrame ' + frame.name)
+            path = self._pdict_list(req_point, 0) #List of points in shortest path
+            velocity = 0
+            for i in range(len(path)-1):
+                velocity += path[i]._pos_dict[path[i+1]].dt(frame) #Sums pos.dt(frame) upto last point
+            return velocity + path[-1].vel(frame) #Adds last point's vel and returns
```

Comment:
You don't seem to set the velocity in `_vel_dict` before returning.

Reply:
You need a unit test that checks that the velocity is returned and that it is set in the dictionary.

Reply:
Yes, I think it should not be updated with calculated velocity because if user changes velocity of any particle in path then the calculated velocity of all affected points should change.

Reply:
We shall not fix  _vel_dict with calculated velocity , this dictionary should be updated by set_vel funct only

Reply:
> Yes, I think it should not be updated with calculated velocity because if user changes velocity of any particle in path then the calculated velocity of all affected points should change.

That is true, but that is expected behavior in this module. It's designed that way, for better or worse. The user is very much responsible for doing operations in a correct order. If they manipulate velocities deep in the chain without updating dependent velocities, then they've screwed up. We don't have a mechanism to correct the velocities in the chain of points given the change in a branch.

Reply:
I'm pretty sure that `ang_vel()` both returns the velocity and sets the velocity. No? We want to copy that behavior.

Reply:
> I'm pretty sure that `ang_vel()` both returns the velocity and sets the velocity. No? We want to copy that behavior.

Yes I'll change that but still I believe not updating _vel_dict would have been better and ang_vel() could  have also been updated in future.

Reply:
`vel()` should return the velocity if it has already been set or try to calculate it, if it can, set it, and then return the set velocity. If you don't set it, then I think in the vast majority of the cases the next line typed by the user will set the velocity. We should copy the `ang_vel()` behavior here for consistency, regardless if it not an optimal design. Note that v1pt and v2pt work this way too.